core/src/main/java/io/parsingdata/metal/data/ConcatenatedValueSource.java

```diff
@@ -79,13 +80,15 @@ private Trampoline<byte[]> getData(final ImmutableList<Value> values, final BigI
         if (length.compareTo(ZERO) <= 0) {
             return complete(() -> output);
         }
-        if (currentOffset.add(values.head.slice().length).compareTo(offset) <= 0) {
-            return intermediate(() -> getData(values.tail, currentOffset.add(values.head.slice().length), currentDest, offset, length, output));
+        final BigInteger nextOffset = currentOffset.add(values.head.slice().length);
+        if (nextOffset.compareTo(offset) <= 0) {
+            return intermediate(() -> getData(values.tail, nextOffset, currentDest, offset, length, output));
         }
         final BigInteger localOffset = offset.subtract(currentOffset).compareTo(ZERO) < 0 ? ZERO : offset.subtract(currentOffset);
-        final BigInteger toCopy = length.compareTo(values.head.slice().length.subtract(localOffset)) > 0 ? values.head.slice().length.subtract(localOffset) : length;
-        System.arraycopy(values.head.slice().getData(), localOffset.intValueExact(), output, currentDest.intValueExact(), toCopy.intValueExact());
```

Comment:
The `values.head.slice().getData()` was the culprit of the slow read. See also [ConcatenatedValueSourceTest](https://github.com/parsingdata/metal/pull/406/files#diff-cd46a952cb6841c5b74fb506794539df74cc0b9571907a8643c61cf8faefdaf7).